qualtran/bloqs/state_preparation/controlled_state_preparation.py

```diff
@@ -0,0 +1,484 @@
+                rom_vals, bb, control, target_state, rot_reg, phase_gradient
+            )
+            control, target_state, rot_reg, phase_gradient = self._prepare_amplitudes(
+                rom_vals, bb, control, target_state, rot_reg, phase_gradient
+            )
+        else:
+            control, target_state, rot_reg, phase_gradient = self._prepare_amplitudes(
+                rom_vals, bb, control, target_state, rot_reg, phase_gradient
+            )
+            control, target_state, rot_reg, phase_gradient = self._prepare_phases(
+                rom_vals, bb, control, target_state, rot_reg, phase_gradient
+            )
+        # deallocate rotation register's qubits
+        qs = bb.split(rot_reg)
+        for q in qs:
+            bb.add(ZeroEffect(), q=q)
+        return {"control": control, "target_state": target_state, "phase_gradient": phase_gradient}
+
+    def _prepare_amplitudes(
+        self,
+        rom_vals: ArrayLike,
+        bb: BloqBuilder,
+        control: SoquetT,
+        target_state: SoquetT,
+        rot_reg: SoquetT,
+        phase_gradient: SoquetT,
+    ):
+        # if it is the adjoint gate, load the modular negative values to undo the rotations that
+        # loaded the amplitudes
+        if self.uncompute:
+            rom_vals = RotationTree.extract_rom_values_from_state(
+                self.state, self.rot_reg_bitsizes, uncompute=True
+            )
+        state_qubits = bb.split(target_state)
+        for i in range(self.state_bitsizes):
+            # for the normal gate loop from qubit 0 to state_bitsizes-1, if it is the adjoint
+            # then the process is run backwards with the opposite turn angles
+            if self.uncompute:
+                qi = self.state_bitsizes - i - 1
+            else:
+                qi = i
+            ctrl_rot_q = ControlledQROMRotateQubit(qi, self.rot_reg_bitsizes, tuple(rom_vals[qi]))
+            state_qubits[qi] = bb.add(Rx(angle=np.pi / 2), q=state_qubits[qi])
+            # first qubit does not have selection registers, only controls
+            if qi == 0:
+                control, state_qubits[qi], rot_reg, phase_gradient = bb.add(
+                    ctrl_rot_q,
+                    prepare_control=control,
+                    qubit=state_qubits[qi],
+                    rot_reg=rot_reg,
+                    phase_gradient=phase_gradient,
+                )
+            else:
+                sel = bb.join(state_qubits[:qi])
+                control, sel, state_qubits[qi], rot_reg, phase_gradient = bb.add(
+                    ctrl_rot_q,
+                    prepare_control=control,
+                    selection=sel,
+                    qubit=state_qubits[qi],
+                    rot_reg=rot_reg,
+                    phase_gradient=phase_gradient,
+                )
+                state_qubits[:qi] = bb.split(sel)
+            state_qubits[qi] = bb.add(Rx(angle=-np.pi / 2), q=state_qubits[qi])
```

Comment:
I see a bunch of code is duplicated mainly to pass the right set of inputs to the bloq. You can rewrite this function in a more succinct way by using the helper `bb.add_d` instead of `bb.add` and using a dictionary to keep track of arguments. For example:

```python
    def _prepare_amplitudes(
        self,
        rom_vals: ArrayLike,
        bb: BloqBuilder,
        control: SoquetT,
        target_state: SoquetT,
        rot_reg: SoquetT,
        phase_gradient: SoquetT,
    ):
        # if it is the adjoint gate, load the modular negative values to undo the rotations that
        # loaded the amplitudes
        if self.uncompute:
            rom_vals = RotationTree(self.state).get_rom_values(
                self.rot_reg_bitsizes, uncompute=True
            )
        state_qubits = bb.split(target_state)
        soqs = {"prepare_control": control, "rot_reg": rot_reg, "phase_gradient": phase_gradient}
        for i in range(self.state_bitsizes):
            # for the normal gate loop from qubit 0 to state_bitsizes-1, if it is the adjoint
            # then the process is run backwards with the opposite turn angles
            if self.uncompute:
                qi = self.state_bitsizes - i - 1
            else:
                qi = i
            ctrl_rot_q = ControlledQROMRotateQubit(qi, self.rot_reg_bitsizes, tuple(rom_vals[qi]))
            state_qubits[qi] = bb.add(Rx(angle=np.pi / 2), q=state_qubits[qi])
            if qi:
                # first qubit does not have selection registers, only controls
                soqs["selection"] = bb.join(state_qubits[:qi])
            soqs = bb.add_d(ctrl_rot_q, **soqs, qubit=state_qubits[qi])
            if qi:
                state_qubits[:qi] = bb.split(soqs.pop("selection"))
            state_qubits[qi] = bb.add(Rx(angle=-np.pi / 2), q=soqs.pop("qubit"))

        target_state = bb.join(state_qubits)
        return soqs["prepare_control"], target_state, soqs["rot_reg"], soqs["phase_gradient"]
```